zephyr/lib/cpu.c

```diff
@@ -143,14 +143,25 @@ int cpu_enable_core(int id)
 		return 0;
 
 #if ZEPHYR_VERSION(3, 0, 99) <= ZEPHYR_VERSION_CODE
-	/* During kernel initialization, the next pm state is set to ACTIVE. By checking this
+	/* Zephyr implementations on secondary cores initialzation are
+	 * different between MTL+ and other platforms.
+	 */
+#if defined(CONFIG_ACE)
+	/* On MTL+ platforms, only run z_init_cpu() on boot, not when waking up from D3,
+	 * regardless whether saving to IMR is enabled or not.
```

Comment:
this is not how I understood the flow in a discussion yesterday. If on MTL+ IMR context saving isn't enabled, then waking up from D3 will be equivalent to the first boot, so the `pm_state_next_get(id)->state == PM_STATE_ACTIVE` condition will be true and `z_init_cpu()` will be run.

Reply:
> this is not how I understood the flow in a discussion yesterday. If on MTL+ IMR context saving isn't enabled, then waking up from D3 will be equivalent to the first boot, so the `pm_state_next_get(id)->state == PM_STATE_ACTIVE` condition will be true and `z_init_cpu()` will be run.

@tmleman Can you help clarify?
I also updated the PR according to your https://github.com/thesofproject/sof/commit/fa1572c4d5668bda0867d46c687808bcdd800458 for mtl-006.

Reply:
On MTL `z_init_cpu(id)` is called for secondary cores only for the first init. As I wrote in the comment in the code:

> During kernel initialization, the next pm state is set to ACTIVE.

If core goes from D0 to D3, next state is set to `PM_STATE_SOFT_OFF`. Thanks to this we know if the core is enabled for the first time.

`CONFIG_ADSP_IMR_CONTEXT_SAVE` changes the behavior of the primary core during D3. If disabled, every exit from D3 for primary core is like first boot. Therefore, if we re-enable the secondary core after that, the function `z_init_cpu(id)` will be called.